src/items.rs

```diff
@@ -232,42 +233,10 @@ impl<'a> FmtVisitor<'a> {
         // Account for sugary self.
         let mut min_args = 1;
         if let Some(explicit_self) = explicit_self {
-            match explicit_self.node {
-                ast::ExplicitSelf_::SelfRegion(ref lt, ref m, _) => {
-                    let lt_str = match lt {
-                        &Some(ref l) => format!("{} ", pprust::lifetime_to_string(l)),
-                        &None => String::new(),
-                    };
-                    let mut_str = match m {
-                        &ast::Mutability::MutMutable => "mut ".to_owned(),
-                        &ast::Mutability::MutImmutable => String::new(),
-                    };
-                    arg_item_strs[0] = format!("&{}{}self", lt_str, mut_str);
-                    min_args = 2;
-                }
-                ast::ExplicitSelf_::SelfExplicit(ref ty, _) => {
-                    arg_item_strs[0] = format!("self: {}", pprust::ty_to_string(ty));
-                }
-                ast::ExplicitSelf_::SelfValue(_) => {
-                    assert!(args.len() >= 1, "&[ast::Arg] shouldn't be empty.");
-
-                    // this hacky solution caused by absence of `Mutability` in `SelfValue`.
-                    let mut_str = {
-                        if let ast::Pat_::PatIdent(ast::BindingMode::BindByValue(mutability), _, _)
-                                = args[0].pat.node {
-                            match mutability {
-                                ast::Mutability::MutMutable => "mut ",
-                                ast::Mutability::MutImmutable => "",
-                            }
-                        } else {
-                            panic!("there is a bug or change in structure of AST, aborting.");
-                        }
-                    };
-
-                    arg_item_strs[0] = format!("{}self", mut_str);
-                    min_args = 2;
-                }
-                _ => {}
+            let self_str = rewrite_explicit_self(explicit_self, args);
+            if self_str.len() > 0 {
+                arg_item_strs[0] = self_str;
+                min_args = 2;
```

Comment:
If I understand correctly, the only change here is that we now change min_args to 2 in the `SelfExplicit` case. Was that the bug before? I expected to see some change with formatting a missing span or something. Could you explain how the bug happened please?